### PR TITLE
fix: harden auth, registry, remote tempdirs, and FFI caches

### PR DESCRIPTION
**Pull request by bb-connor**
## Summary
This PR implements and validates the Rust audit findings across control-api, registry, hush-cli/hushd, and native bridges.

### Fixed issues
- Enforced JWT issuer/audience binding in control-api and added claim validation tests.
- Secured remote-extends temp git dir creation in hush-cli and hushd via atomic tempfile dirs.
- Replaced unchecked i64<->u64 DB casts in registry with checked conversions and integrity errors.
- Stopped swallowing non-duplicate-column migration errors in registry schema migration path.
- Fixed registry blob store idempotent concurrent write race.
- Bounded watermarker caches in Go/Python native bridges and added cache bound tests.
- Added OIDC algorithm allowlist enforcement in registry validator.
- Corrected grant exercise validation ordering in control-api.
- Removed lock-held-across-await in hushd alerting exporter shutdown path.

### Validation
- cargo fmt --all
- cargo check --workspace
- cargo clippy --workspace --all-targets --all-features -- -D warnings
- cargo test --workspace
- Targeted verification also run for touched crates (clawdstrike-control-api, clawdstrike-registry, hush-cli, hushd, hush-native).

### Notes
- crates/bridges/hush-go-native is outside root workspace membership. It was validated in an isolated cargo run with equivalent dependencies and passes, and this PR makes the required clawdstrike full feature explicit for standalone builds.

---

> [!NOTE]
> **Medium Risk**
> Touches authentication/token-validation and registry persistence paths (JWT issuer/audience binding, OIDC algorithm allowlist, DB type conversions/migrations, concurrent blob writes), so misconfiguration or edge cases could reject valid traffic or surface new errors despite added tests.
> 
> **Overview**
> **Hardens authentication and token validation across services.** Control API JWTs now require matching `iss`/`aud` (new `JWT_ISSUER`/`JWT_AUDIENCE` config + Helm/docker/docs updates) and the `fleet-fixture` script emits tokens with those claims; registry OIDC validation now enforces a per-provider signing-algorithm allowlist.
> 
> **Makes registry/storage and runtime helpers more defensive.** Registry migrations no longer swallow non-duplicate-column errors, SQLite `i64`↔`u64` conversions are checked (reject negative/out-of-range values), and blob writes are made idempotent under concurrent writers. Remote policy `git` temp directories in `hush-cli`/`hushd` switch to `tempfile`-managed dirs, native Go/Python watermarker caches are bounded (LRU-style), Control API grant exercise validates request/event ordering, and `hushd` alerting exporter shutdown avoids holding a lock across awaits (with new regression tests).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4b3f2bce3da7d7bc42b90720646d6517ebe705f6. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: crates/services/clawdstrike-registry/src/error.rs [hunk 1]
```diff
@@ -21,7 +21,7 @@ pub enum RegistryError {
     Unauthorized(String),
 
     #[error("database error: {0}")]
-    Database(#[from] rusqlite::Error),
+    Database(rusqlite::Error),
```
[the remark this answers is not in the document]
Comment (bb-connor) [reply crates/services/clawdstrike-registry/src/error.rs]: Fixed in 30e333c1e: added #[source] to RegistryError::Database(rusqlite::Error) so error chaining remains intact while preserving the custom From<rusqlite::Error> mapping.

File: crates/bridges/hush-go-native/src/lib.rs [hunk 2]
```diff
@@ -391,9 +392,51 @@ pub unsafe extern "C" fn hush_sanitize_output(
...
+        self.order.push_back(key.clone());
+        self.entries.insert(key, value);
+    }
+
+    fn touch(&mut self, key: &str) {
+        if let Some(idx) = self.order.iter().position(|entry| entry == key) {
+            self.order.remove(idx);
+        }
+        self.order.push_back(key.to_string());
+    }
```
[the remark this answers is not in the document]
Comment (bb-connor) [reply crates/bridges/hush-go-native/src/lib.rs:436]: Good callout. I am deferring cross-crate extraction in this hardening PR because the Go bridge is intentionally outside workspace membership and the Python bridge has a different build surface; I wanted to avoid introducing packaging/build coupling while shipping the security/reliability fixes. I kept parity by adding bounded-cache tests in both bridges. I will track shared extraction as a separate refactor.

**Comment by bb-connor** [comment]
@codex

File: crates/services/clawdstrike-registry/src/error.rs [hunk 3]
```diff
@@ -33,6 +33,17 @@ pub enum RegistryError {
...
+impl From<rusqlite::Error> for RegistryError {
+    fn from(err: rusqlite::Error) -> Self {
+        match err {
+            rusqlite::Error::FromSqlConversionFailure(_, _, source) => {
+                RegistryError::Integrity(source.to_string())
+            }
+            other => RegistryError::Database(other),
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (bb-connor) [reply crates/services/clawdstrike-registry/src/error.rs]: Fixed in d04e0cc39. I removed the special-case FromSqlConversionFailure -> RegistryError::Integrity mapping and now convert rusqlite errors via RegistryError::Database (HTTP 500). This keeps data-corruption/conversion faults as server-side errors rather than 400 client errors. Updated affected registry DB tests accordingly.

File: packages/sdk/hush-py/hush-native/src/lib.rs [hunk 4]
```diff
@@ -234,9 +234,51 @@ fn sanitize_output_native(
...
+        self.order.push_back(key.clone());
+        self.entries.insert(key, value);
+    }
+
+    fn touch(&mut self, key: &str) {
+        if let Some(idx) = self.order.iter().position(|entry| entry == key) {
+            self.order.remove(idx);
+        }
+        self.order.push_back(key.to_string());
+    }
```
[the remark this answers is not in the document]
Comment (bb-connor) [reply packages/sdk/hush-py/hush-native/src/lib.rs:278]: Acked. Same rationale as the parallel thread: I am intentionally deferring extraction of a shared WatermarkerCache helper to a follow-up refactor so this PR stays focused on security/correctness hardening and avoids introducing new cross-crate build coupling between the Go and Python native bridges. Both implementations are kept in lockstep here and each has bounded-cache tests.

File: crates/services/control-api/src/auth/jwt.rs [hunk 5]
```diff
@@ -13,22 +13,38 @@ pub struct Claims {
     pub sub: Uuid,
     pub tenant_id: Uuid,
     pub role: String,
+    pub iss: String,
+    pub aud: String,
```
[the remark this answers is not in the document]
Comment (bb-connor) [reply crates/services/control-api/src/auth/jwt.rs]: Fixed in 294526b42. `Claims.aud` now uses an untagged enum that accepts both single-string and array-form JWT audiences, and I added a regression test (`decode_claims_accepts_array_audience_containing_expected_value`) to cover multi-audience tokens that include the configured audience.

**Comment by bb-connor** [comment]
@codex

File: crates/services/clawdstrike-registry/src/storage.rs [hunk 6]
```diff
@@ -108,14 +106,28 @@ impl BlobStorage {
...
 
+    fn write_blob_if_missing(&self, blob_path: &Path, data: &[u8]) -> Result<(), RegistryError> {
+        let tmp_path = self.root.join(format!(".tmp-{}", uuid::Uuid::new_v4()));
+        fs::write(&tmp_path, data)?;
+        match fs::rename(&tmp_path, blob_path) {
+            Ok(()) => Ok(()),
+            Err(err) if err.kind() == std::io::ErrorKind::AlreadyExists => {
+                let _ = fs::remove_file(&tmp_path);
+                Ok(())
+            }
```
[the remark this answers is not in the document]
Comment (bb-connor) [reply crates/services/clawdstrike-registry/src/storage.rs]: Fixed in 41bd1f4ce. I removed the `AlreadyExists`-specific `rename` arm and replaced it with a single error path that reconciles idempotent concurrent writers by checking whether the destination hash path now exists after a rename failure. This keeps the race-safe behavior while removing the misleading dead branch.

File: crates/services/clawdstrike-registry/src/db.rs [hunk 7]
```diff
@@ -173,6 +174,54 @@ fn escape_sqlite_like_literal(value: &str) -> String {
...
+        Err(err) if is_duplicate_column_error(&err) => Ok(()),
+        Err(err) => Err(err.into()),
+    }
+}
+
+fn sqlite_i64_to_u64(value: i64, field: &str) -> Result<u64, RegistryError> {
+    u64::try_from(value).map_err(|_| {
+        RegistryError::Integrity(format!("invalid negative value for {field}: {value}"))
+    })
+}
```
[the remark this answers is not in the document]
Comment (bb-connor) [reply crates/services/clawdstrike-registry/src/db.rs:205]: Fixed in 4b3f2bce3. `sqlite_i64_to_u64` now maps negative conversion faults to `RegistryError::Database` using `rusqlite::Error::FromSqlConversionFailure`, matching the row-mapper behavior (HTTP 500 server error path). I also added a regression unit test (`sqlite_i64_to_u64_maps_negative_values_to_database_error`).
